### PR TITLE
Tickets/dm 10308 Support rectangular regions in bright object masks

### PR DESCRIPTION
**Pull request by RobertLuptonTheGood**
I've tested the pieces using the new mask file that Jean sent, but not end-to-end.

### CODE REVIEW

File: python/lsst/pipe/tasks/ingest.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ class ParseConfig(Config):
                             doc="Properties and name of translator method")
     defaults = DictField(keytype=str, itemtype=str, default={},
                          doc="Default values if header is not present")
-    hdu = Field(dtype=int, default=0, doc="HDU to read for metadata")
+    hdu = Field(dtype=int, default=-(1 << 31), doc="HDU to read for metadata") # default == INT_MIN
```
Comment (PaulPrice) [inline python/lsst/pipe/tasks/ingest.py]: Is this assuming something about `sizeof(int)`?  Can you get the default value out of C++-land?

File: python/lsst/pipe/tasks/assembleCoadd.py [hunk 2]
```diff
@@ -753,9 +753,20 @@ def setBrightObjectMasks(self, exposure, dataId, brightObjectMasks):
 
         for rec in brightObjectMasks:
             center = afwGeom.PointI(wcs.skyToPixel(rec.getCoord()))
-            radius = rec["radius"].asArcseconds()/plateScale   # convert to pixels
+            if rec["type"] == "box":
+                width = rec["width"].asArcseconds()/plateScale   # convert to pixels
+                height = rec["height"].asArcseconds()/plateScale # convert to pixels
```
Comment (PaulPrice) [inline python/lsst/pipe/tasks/assembleCoadd.py]: Nitpick: Supposed to be two spaces before `#`.

File: python/lsst/pipe/tasks/assembleCoadd.py [hunk 3]
```diff
@@ -753,9 +753,20 @@ def setBrightObjectMasks(self, exposure, dataId, brightObjectMasks):
...
+
+                bbox = afwGeom.BoxI(afwGeom.PointI(int(center[0] - 0.5*width), int(center[1] - 0.5*height)),
+                                    afwGeom.PointI(int(center[0] + 0.5*width), int(center[1] + 0.5*height)))
+                foot = afwDetect.Footprint(bbox, exposure.getBBox())
```
Comment (PaulPrice) [inline python/lsst/pipe/tasks/assembleCoadd.py]: Add a comment about rotation?

Comment (PaulPrice) [inline python/lsst/pipe/tasks/assembleCoadd.py]: Maybe cleaner:

    halfWidth = afwGeom.ExtentI(0.5*afwGeom.ExtentD(0.5*width, 0.5*height))
    bbox = afwGeom.Box2I(center - halfWidth, center + halfWidth)

File: python/lsst/pipe/tasks/objectMasks.py [hunk 4]
```diff
@@ -60,12 +64,18 @@ def readFits(fileName, hdu=0, flags=0):
 
         wcs; fk5
 
-        circle(RA, DEC, RADIUS) # ID: 1
+        circle(RA, DEC, RADIUS)           # ID: 1, mag: 12.34
+        box(RA, DEC, XSIZE, YSIZE, THETA) # ID: 2, mag: 23.45
+
+        The ", mag: XX.YY" is optional
+
+        All
```
Comment (PaulPrice) [inline python/lsst/pipe/tasks/objectMasks.py]: Something missing here.

File: python/lsst/pipe/tasks/objectMasks.py [hunk 5]
```diff
@@ -60,12 +64,18 @@ def readFits(fileName, hdu=0, flags=0):
...
+        All 
 
         The commented lines must be present, with the relevant fields such as tract patch and filter filled
-        in. The coordinate system must be listed as above. Each patch is specified as a circle, with an RA,
-        DEC, and Radius specified in decimal degrees. Only circles are supported as region definitions
-        currently.
+        in. The coordinate system must be listed as above. Each patch is specified as a box or circle, with
+        RA, DEC, and dimensions specified in decimal degrees (with or without an explicit "d")
+
+        Only (axis-aligned) boxes and circles are currently supported as region definitions
```
Comment (PaulPrice) [inline python/lsst/pipe/tasks/objectMasks.py]: Nitpick: add full-stops for all paragraphs.

File: python/lsst/pipe/tasks/objectMasks.py [hunk 6]
```diff
@@ -104,24 +114,52 @@ def readFits(fileName, hdu=0, flags=0):
...
                 # This regular expression parses the regions file for each region to be masked,
                 # with the format as specified in the above docstring.
-                mat = re.search(r"^\s*circle(?:\s+|\s*\(\s*)"
+                mat = re.search(r"^\s*(box|circle)(?:\s+|\s*\(\s*)"
                                 "(\d+(?:\.\d*)([d]*))" "(?:\s+|\s*,\s*)"
-                                "([+-]?\d+(?:\.\d*)([d]*))" "(?:\s+|\s*,\s*)"
-                                "(\d+(?:\.\d*))([d'\"]*)" "(?:\s*|\s*\)\s*)"
-                                "\s*#\s*ID:\s*(\d+)" "\s*$", line)
+                                "([+-]?\d+(?:\.\d*))([d]*)" "(?:\s+|\s*,\s*)"
+                                "([+-]?\d+(?:\.\d*))([d]*)" "(?:\s+|\s*,\s*)?"
```
Comment (PaulPrice) [inline python/lsst/pipe/tasks/objectMasks.py]: Does this now allow `circle` with an extra (nonsense) parameter?

File: python/lsst/pipe/tasks/objectMasks.py [hunk 7]
```diff
@@ -104,24 +114,52 @@ def readFits(fileName, hdu=0, flags=0):
...
+                                "(?:(\d+(?:\.\d*))([d'\"]*)\s*,\s*(\d+)(?:\.\d*([d]*))?)?"
+                                "(?:\s*|\s*\)\s*)"   # close paren
+                                "\s*#\s*ID:\s*(\d+)" # start comment
+                                "(?:\s*,\s*mag:\s*(\d+\.\d*))?"
+                                "\s*$", line)
                 if mat:
-                    ra, raUnit, dec, decUnit, radius, radiusUnit, _id = mat.groups()
+                    _type, ra, raUnit, dec, decUnit, \
+                        param1, param1Unit, param2, param2Unit, param3, param3Unit, \
+                        _id, mag = mat.groups()
```
Comment (PaulPrice) [inline python/lsst/pipe/tasks/objectMasks.py]: That's a bit ugly.  Would parens remove the need for backslashes here?

File: python/lsst/pipe/tasks/objectMasks.py [hunk 8]
```diff
@@ -104,24 +114,52 @@ def readFits(fileName, hdu=0, flags=0):
...
                     dec = convertToAngle(dec, decUnit, "dec", fileName, lineNo)
-                    radius = convertToAngle(radius, radiusUnit, "radius", fileName, lineNo)
+
+                    if _type == "box":
+                        width = convertToAngle(param1, param1Unit, "xsize", fileName, lineNo)
+                        height = convertToAngle(param2, param2Unit, "ysize", fileName, lineNo)
+                        angle = convertToAngle(param3, param3Unit, "theta", fileName, lineNo)
+
+                        if angle != 0.0:
+                            log.warn("Rotated boxes are not supported: \"%s\" at %s:%d" % (
```
Comment (PaulPrice) [inline python/lsst/pipe/tasks/objectMasks.py]: Maybe an exception might be appropriate for now?

File: python/lsst/pipe/tasks/objectMasks.py [hunk 9]
```diff
@@ -104,24 +114,52 @@ def readFits(fileName, hdu=0, flags=0):
...
                     rec["id"] = _id
+                    rec["mag"] = mag
                     rec.setCoord(afwCoord.Fk5Coord(ra, dec))
-                    rec["radius"] = radius
+
+                    if _type == "box":
+                        rec["height"] = height
+                        rec["width"] = width
+                    elif _type == "circle":
+                        rec["radius"] = radius
```
Comment (PaulPrice) [inline python/lsst/pipe/tasks/objectMasks.py]: Dummy values (`NaN`, perhaps) for the type-inappropriate columns?

File: python/lsst/pipe/tasks/assembleCoadd.py (hunk 3 above)
Comment (RobertLuptonTheGood) [inline python/lsst/pipe/tasks/assembleCoadd.py]: Good idea, although that code snippet is pretty buggy...

File: python/lsst/pipe/tasks/objectMasks.py (hunk 6 above)
Comment (RobertLuptonTheGood) [inline python/lsst/pipe/tasks/objectMasks.py]: Fixed.

File: python/lsst/pipe/tasks/objectMasks.py (hunk 8 above)
Comment (RobertLuptonTheGood) [inline python/lsst/pipe/tasks/objectMasks.py]: I didn't want an exception as it'd fail to keep checking the file.  Added a counter, so I now throw 
the exception after the entire file has been parsed.

File: python/lsst/pipe/tasks/objectMasks.py (hunk 9 above)
Comment (RobertLuptonTheGood) [inline python/lsst/pipe/tasks/objectMasks.py]: Experimentally afwTable does that, but I've made it explicit.  Removes a second if block looking
at region types too.

**Pull request closed without merging**